library/src/image.cxx

```diff
@@ -121,63 +122,89 @@ image& image::operator=(image&& img) noexcept
 //----------------------------------------------------------------------------
 unsigned int image::getWidth() const
 {
-  return this->Internals->Width;
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  return dims[0];
 }
 
 //----------------------------------------------------------------------------
 unsigned int image::getHeight() const
 {
-  return this->Internals->Height;
+  int dims[3];
+  this->Internals->Image->GetDimensions(dims);
+  return dims[1];
 }
 
+#ifndef F3D_NO_DEPRECATED
 //----------------------------------------------------------------------------
 image& image::setResolution(unsigned int width, unsigned int height)
 {
-  this->Internals->Width = width;
-  this->Internals->Height = height;
-  this->Internals->UpdateBufferSize();
+  this->Internals->Image->SetDimensions(static_cast<int>(width), static_cast<int>(height), 1);
+  this->Internals->Image->AllocateScalars(
+    VTK_UNSIGNED_CHAR, static_cast<int>(this->getChannelCount()));
   return *this;
 }
+#endif
 
 //----------------------------------------------------------------------------
 unsigned int image::getChannelCount() const
 {
-  return this->Internals->Channels;
+  return this->Internals->Image->GetNumberOfScalarComponents();
 }
 
+#ifndef F3D_NO_DEPRECATED
 //----------------------------------------------------------------------------
 image& image::setChannelCount(unsigned int dim)
 {
-  this->Internals->Channels = dim;
-  this->Internals->UpdateBufferSize();
+  this->Internals->Image->AllocateScalars(VTK_UNSIGNED_CHAR, static_cast<int>(dim));
   return *this;
 }
+#endif
+
+//----------------------------------------------------------------------------
+image::ChannelType image::getChannelType() const
+{
+  switch (this->Internals->Image->GetScalarType())
+  {
+    case VTK_UNSIGNED_CHAR:
+      return ChannelType::BYTE;
+    case VTK_UNSIGNED_SHORT:
+      return ChannelType::SHORT;
+    case VTK_FLOAT:
+      return ChannelType::FLOAT;
+    default:
+      break;
+  }
+  throw read_exception("Unknown channel type");
+}
 
 //----------------------------------------------------------------------------
-image& image::setData(unsigned char* buffer)
+image& image::setData(void* buffer)
```

Comment:
why changing to void* then casting to unsigned char* later ?

Reply:
Because std::copy needs a type.

Reply:
see other discussions